pulp_rpm/app/tasks/synchronizing.py

```diff
@@ -1583,31 +1583,40 @@ async def run(self):
                     content_q_by_type[model_type] = content_q_by_type[model_type] | unit_q
                     d_content_by_nat_key[d_content.content.natural_key()].append(d_content)
 
-            for model_type, content_q in content_q_by_type.items():
-                try:
-                    await sync_to_async(model_type.objects.filter(content_q).touch)()
-                except AttributeError:
-                    raise TypeError(
-                        "Plugins which declare custom ORM managers on their content classes "
-                        "should have those managers inherit from "
-                        "pulpcore.plugin.models.ContentManager."
-                    )
+            # For each entry in the batch, determine if we need to "fix up" the data before
+            # letting it pass down the pipeline.
+            #
+            # NOTE that we "remember" entities we've touched, and **which fields** we've touched,
+            # in order to persist **just the changes we make** below. This helps us avoid some
+            # nasty deadlock-windows between save() and touch() in this stage.
+            #
+            # IF/WHEN/AS YOU ADD TO THIS CODE - make sure you update modified_rslts_by_pk and
+            # modified_fields_by_pk, or your changes WILL NOT BE PERSISTED.
+
+            # holds the to-be-saved results by-pk
+            modified_rslts_by_pk = {}
+            # per to-be-saved pk, holds a list of field-names of the things we've changed'
+            modified_fields_by_pk = defaultdict(list)
+            query_types = content_q_by_type.keys()
+            for model_type in query_types:
                 async for result in sync_to_async_iterable(
-                    model_type.objects.filter(content_q).iterator()
+                    model_type.objects.filter(content_q_by_type[model_type]).iterator()
                 ):
                     for d_content in d_content_by_nat_key[result.natural_key()]:
                         # ============ The below lines are added vs. pulpcore ============
```

Comment:
Can we demarcate what we're changing here as well?  This just helps keep it straight between core and RPM